datafusion/common/src/scalar/mod.rs

```diff
@@ -854,6 +854,140 @@ pub fn get_dict_value<K: ArrowDictionaryKeyType>(
     Ok((dict_array.values(), dict_array.key(index)))
 }
 
+/// Cache for dictionary key arrays to avoid repeated allocations
+/// when the same size is used frequently.
+///
+/// Similar to PartitionColumnProjector's ZeroBufferGenerators, this cache
+/// stores key arrays for different dictionary key types. The cache is
+/// limited to 1 entry per type (the last size used).
+#[derive(Debug)]
+struct KeyArrayCache<K: ArrowDictionaryKeyType> {
+    cache: Option<(usize, bool, PrimitiveArray<K>)>, // (size, is_null, key_array)
+}
+
+impl<K: ArrowDictionaryKeyType> Default for KeyArrayCache<K> {
+    fn default() -> Self {
+        Self { cache: None }
+    }
+}
+
+impl<K: ArrowDictionaryKeyType> KeyArrayCache<K> {
+    fn get_or_create(&mut self, size: usize, is_null: bool) -> PrimitiveArray<K> {
```

Comment:
nit is I would use `num_rows` here rather than size as the size might be confused with the key size (eg `Int16Type` would have size 16??)